oonib/testhelpers/http_helpers.py

```diff
@@ -1,16 +1,87 @@
+import os
 import json
 import random
 import string
+import tempfile
+from hashlib import sha256
 
-from twisted.internet import protocol, defer
+from urlparse import urlparse
 
-from cyclone.web import RequestHandler, Application
+from twisted.python.failure import Failure
+
+from twisted.internet.protocol import Factory, Protocol
+from twisted.internet.endpoints import TCP4ClientEndpoint
+
+from twisted.internet.error import DNSLookupError, TimeoutError
+from twisted.internet.error import ConnectionRefusedError
+from twisted.internet import protocol, defer, reactor
+
+from twisted.names.error import DNSNameError, DNSServerError
+from twisted.names import client as dns_client
+from twisted.names import dns
+
+from twisted.web._newclient import HTTPClientParser, ParseError
+from twisted.web.client import Agent, BrowserLikeRedirectAgent, readBody
+from twisted.web.client import ContentDecoderAgent, GzipDecoder
+from twisted.web.http_headers import Headers
+from twisted.web import error
+
+from twisted.web._newclient import ResponseFailed
+
+from cyclone.web import RequestHandler, Application, HTTPError
+from cyclone.web import asynchronous
 
 from twisted.protocols import policies, basic
 from twisted.web.http import Request
 
 from oonib import log, randomStr
 
+# Monkey patch to overcome issue in parsing of HTTP responses that don't
+# contain 3 parts in the HTTP response line
```

Comment:
It happens because some people decided to write their own HTTP server that doesn't fully respect the HTTP spec. An example of this okcupid.com:

```
curl -kis okcupid.com

HTTP/1.1 302
Date: Tue, 24 May 2016 10:40:50 GMT
Transfer-Encoding: chunked
Connection: keep-alive
Set-Cookie: __cfduid=dda563ef0b63790ff641999bbb5432a231464086450; expires=Wed, 24-May-17 10:40:50 GMT; path=/; domain=.okcupid.com; HttpOnly
X-OKWS-Version: OKWS/3.1.38.0
```

What's weird is that from the source code they have published it looks like it actually does support setting the status code reason string (https://github.com/okws/okws/blob/80f41cc0ff770a83e9f9e468cfdc2040dd7e0c43/libahttp/resp.h#L55), but in their deployment it doesn't. Probably this is due to some other middle box mangling with the response line.